pepr/privacy/mia.py

```diff
@@ -0,0 +1,1182 @@
+        self.attack_results = Mia._attack_model_evaluation(
+            attack_models,
+            target_prediction,
+            target_data,
+        )
+        logger.info("Calculate shadow model summary.")
+        self.attack_results.update(
+            Mia._shadow_model_evaluation(
+                shadow_models,
+                shadow_data_indices,
+                shadow_train_data,
+                shadow_train_labels,
+            )
+        )
+
+        # Print attack summary
+        def _list_to_formatted_string(arr):
+            string = ""
+            for item in arr:
+                string = string + f"{round(item, 3):>10}"
+            return string
+
+        logger.info(
+            "Attack Summary"
+            f"\n"
+            f"\n###################### Shadow Results ######################"
+            f"\n"
+            + f"\n{'Shadow Models:':<30}"
+            + f"\n{'Average Training Accuracy:':<30}"
+            + _list_to_formatted_string(self.attack_results["shadow_train_accuracy"])
+            + f"\n{'Average Evaluation Accuracy:':<30}"
+            + _list_to_formatted_string(self.attack_results["shadow_test_accuracy"])
+            + f"\n"
+            f"\n###################### Attack Results ######################"
+            f"\n"
+            f"\n{'Attack Models:':<30}"
+            + f"\n{'Average Test Accuracy:':<30}"
+            + _list_to_formatted_string([self.attack_results["overall_test_accuracy"]])
+            + f"\n{'Average Precision:':<30}"
+            + _list_to_formatted_string([self.attack_results["overall_precision"]])
+            + f"\n{'Average Recall:':<30}"
+            + _list_to_formatted_string([self.attack_results["overall_recall"]])
+        )
+
+    @staticmethod
+    def _get_target_model_indices(
+        target_indices,
+        evaluation_indices,
+        record_indices_per_target,
+        attack_test_labels,
+        number_classes,
+    ):
+        """
+        Calculate classified indices per target with their true attack label for
+        evaluation and convert indices to evaluation dataset space.
+
+        Parameters
+        ----------
+        target_indices : list
+            List of indices used to train all target models. Indices in complete
+            dataset range.
+        evaluation_indices : list
+            List of indices for attack model evaluation. Indices in complete dataset
+            range.
+        record_indices_per_target : numpy.ndarray
+            List of mapping which records of the target_indices are used to train a
+            target model.
+        attack_test_labels : numpy.ndarray
+            List of labels in the evaluation dataset.
+        number_classes : int
+            Number of classes in the dataset.
+
+        Returns
+        -------
+        dict
+            Dictionary storing the classified indices and labels. Index range is the
+            evaluation data segment.
```

Comment:
Can we refine what's meant by the evaluation data segment? :)

Reply:
Would "data slice" or "data partition" be better? The sentence means, that the indices point to elements in the evaluation data array, not to elements in the full dataset, which are also in the evaluation dataset.

Reply:
Yes, I think `data slice` would be better. And probably you can add the explanation, that the indices point to elements in the evaluation data array.  I think then it should be clear :)